cpp/tests/io/text/data_chunk_source_test.cpp

```diff
@@ -114,4 +115,215 @@ TEST_F(DataChunkSourceTest, Host)
+                                             0,   // field length
+                                             1,  2,
+                                             56,  // magic number
+                                             78,  // magic number
+                                             1,   // field length
+                                             0,   // field length
+                                             3,   //
+                                             90,  // magic number
+                                             12,  // magic number
+                                             8,   // field length
+                                             0,   // field length
+                                             1,  2, 3, 4, 5, 6, 7, 8}};
+  stream.write(reinterpret_cast<const char*>(header.data()), header.size());
+  uint16_t extra_size = extra_blocklen_field.size() + 2;
+  if (add_extra_garbage_before) { extra_size += extra_garbage_field1.size(); }
+  if (add_extra_garbage_after) { extra_size += extra_garbage_field2.size(); }
+  write_int(stream, extra_size);
+  if (add_extra_garbage_before) {
+    stream.write(extra_garbage_field1.data(), extra_garbage_field1.size());
+  }
+  stream.write(extra_blocklen_field.data(), extra_blocklen_field.size());
+  auto compressed_size          = data.size() + 5;
+  uint16_t block_size_minus_one = compressed_size + 19 + extra_size;
+  write_int(stream, block_size_minus_one);
+  if (add_extra_garbage_after) {
+    stream.write(extra_garbage_field2.data(), extra_garbage_field2.size());
+  }
+  write_int<uint8_t>(stream, 1);
+  write_int<uint16_t>(stream, data.size());
+  write_int<uint16_t>(stream, ~static_cast<uint16_t>(data.size()));
+  stream.write(data.data(), data.size());
+  // this does not produce a valid file, since we write 0 as the CRC
+  // the parser ignores the checksum, so it doesn't matter to the test
+  // to check output with gzip, plug in the CRC of `data` here.
+  write_int<uint32_t>(stream, 0);
+  write_int<uint32_t>(stream, data.size());
+}
+
+void write_bgzip(std::ostream& stream,
+                 const std::string& data,
+                 std::default_random_engine& rng,
+                 bool write_eof = true)
+{
+  // make sure the block size with header stays below 65536
+  std::uniform_int_distribution<std::size_t> block_size_dist{1, 65000};
+  auto begin     = data.begin();
+  auto const end = data.end();
+  int i          = 0;
+  while (begin < end) {
+    auto len = std::min<std::size_t>(end - begin, block_size_dist(rng));
+    write_bgzip_block(stream, std::string{begin, begin + len}, i & 1, i & 2);
+    begin += len;
+    i++;
+  }
+  if (write_eof) { write_bgzip_block(stream, {}, false, false); }
+}
+
+TEST_F(DataChunkSourceTest, BgzipSource)
+{
+  auto filename = temp_env->get_temp_filepath("bgzip_source");
+  std::string input{"bananarama"};
+  for (int i = 0; i < 24; i++) {
+    input = input + input;
+  }
```

Comment:
if I'm reading this right, input adds up to 160MB?

Reply:
correct, I wanted something that is larger than the default read size `chunk_load_size`, which is 16 MB